List possible overload variants when none match

Michael0x2a opened the pull request:
Resolves https://github.com/python/mypy/issues/672 and supersedes https://github.com/python/mypy/pull/5054.

Currently, if a user calls an overloaded function and mypy is unable to find a match, mypy will display display a warning like 'No overload variant of "func" matches argument types" with no additional context.

This pull request will list several matching variants in addition to that message. If possible, mypy will attempt to show only overloads that have a compatible number of arguments. The number of overloads shown is always capped at a maximum of 2.

This pull request does *not* attempt to report a custom error when...

1. Union-math fails. Rationale: the pending PR will change how unions are handled dramatically enough to the point where any error handling here would be pointless.

2. Mypy is able to infer what signature the user most likely meant by looking at the erased types. Rationale: I attempted to implement this feature but was unable to make it consistently work without adding several ugly hacks around how mypy records errors so decided to defer implementing this feature.

ilevkivskyi commented:
Sorry, it looks like I will not be able to review today :-( Probably on Saturday evening.

Michael0x2a commented:
@ilevkivskyi -- no worries, take your time :)

If you also want to prioritize landing the other overload PRs first, I'm fine with that. This PR feels like the kind of thing that'll cause a bunch of merge conflicts, so maybe it makes sense to just defer reviewing this so I can fix the conflicts all in one go instead of distributing them through the other PRs.

Michael0x2a commented:
Up to you though, of course.

Michael0x2a commented:
@gvanrossum -- before I merge this in, did you want to take a quick look? I know you said that the error messages were too noisy in the previous incarnation of this PR. If you still feel that way, there are probably one or two things I could try doing to make the error message shorter.

gvanrossum commented on mypy/messages.py:
```diff
@@ -1233,13 +1243,17 @@ def report_protocol_problems(self, subtype: Union[Instance, TupleType, TypedDict
+        shown = min(max_items, len(targets))
+        max_available = len(func.items())
+        if shown < max_available:
+            self.note('<{} more overload(s) not shown>'.format(max_available - shown),
```
Apart form the uglines around plurals, it seems there are two possible reasons why not all overloads are shown:

- because they don't match the number of arguments (or the keywords or whatever)
- because the sheer number exceeds `max_items`

I think it would be good to differentiate between these in the messages somehow.

gvanrossum commented on mypy/messages.py:
```diff
@@ -1233,13 +1243,17 @@ def report_protocol_problems(self, subtype: Union[Instance, TupleType, TypedDict
+        if not targets:
+            targets = func.items()
+        for item in targets[:max_items]:
             self.note('@overload', context, offset=2 * offset)
```
The extra line with `@overload` just doubles the verbosity without adding any information. I think it's okay to leave it out.

Michael0x2a commented:
@gvanrossum and/or @ilevkivskyi -- how's this?

The new format should handle plurals correctly and will display a message that looks like this if there are more then 2 overloads that have matching types/arity/whatever:

    Possible overload variants:
        def foo(x: int) -> int
        def foo(x: str) -> str
        <2 more matching overloads not shown, out of 5 total overloads>

...or like this if we're suppressing only non-matching overloads:

    Possible overload variants:
        def foo(x: int) -> int
        def foo(x: str) -> str
        <1 more non-matching overload not shown>

...and of course, like this if we we were able to display all the variants:

    Possible overload variants:
        def foo(x: int) -> int
        def foo(x: str) -> str

(I guess since we cut the number of lines shown in half by removing `@overload`, we could also increase the max number of variants shown to 3 or 4 or something. I don't particularly care either way though.)

ilevkivskyi commented on mypy/messages.py:
```diff
@@ -1233,14 +1238,49 @@ def report_protocol_problems(self, subtype: Union[Instance, TupleType, TypedDict
+        if shown < max_available:
+            left = max_available - shown
+            msg = '<{} more overload{} not shown>'.format(left, plural_s(left))
+            self.note(msg, context, offset=2 * offset)
```
All changes in this function are unnecessary now, except `plural_s` usage.

ilevkivskyi commented on mypy/messages.py:
```diff
@@ -1233,14 +1238,49 @@ def report_protocol_problems(self, subtype: Union[Instance, TupleType, TypedDict
+        assert shown <= max_matching <= max_available
+        if shown < max_matching <= max_available:
+            left = max_matching - shown
+            msg = '<{} more matching overload{} not shown, out of {} total overloads>'.format(
```
I would rather say this as `{} more matching and {} non-matching overload{} not shown`. But this is still not perfect because the overloads aren't "matching" (as the error message above says). Anyway, I think it is not so important to delay merging this PR, so I will go ahead unless someone proposes better wording.

ilevkivskyi commented:
@Michael0x2a Will you have time to make minor final updates here before I cut the release branch tomorrow morning? (If not it is also OK.)

Michael0x2a commented:
@ilevkivskyi -- most likely. I'm planning on leaving work + finishing dinner in roughly 5 hours or so and can start making any necessary changes after that.

I guess the timing might be a little tight, depending on whether you mean morning Dublin time vs Pacific time, but eh, whatever.

Michael0x2a commented:
@gvanrossum and @ilevkivskyi  -- Mkay, I reverted the changes to `pretty_overload` (apart from some tweaks to make pluralization tidier), swapped out the word "matching" for "similar" for the first error message, and added the tweak Guido suggested (it ended up being pretty simple to implement in the end).

I'll merge this in once the tests pass (unless there are any last-minute objections, I guess).
